File: calculator/fifo.js
// Simulación del algoritmo de reemplazo de páginas FIFO - Versión JS

// Secuencia de referencias de páginas
const referencias = ["X", "Y", "Z", "W", "X", "Y", "L", "X", "Y", "Z", "W", "L"];
const marcos = 4; // Número de marcos disponibles

// Inicialización
let memoria = []; // Lista de páginas en memoria
let cola_fifo = []; // Cola FIFO para el orden de llegada
let fallos = []; // Lista de "SI" o "NO" indicando si hubo fallo de página
let estados_memoria = []; // Lista para guardar el estado de los marcos en cada instante

console.log("=== SIMULACIÓN PASO A PASO ===");
let header = "Instante | Referencia ";
for (let i = 0; i < marcos; i++) header += `| Marco ${i} `;
header += "| Fallo | Acción";
console.log(header);
console.log("-".repeat(header.length));

// Simulación de FIFO
for (let i = 0; i < referencias.length; i++) {
    const pagina = referencias[i];
    let accion = "";
    if (memoria.includes(pagina)) {
        fallos.push("X"); // Hit
        accion = "Hit - página ya en memoria";
    } else {
        fallos.push("F"); // Fallo
        if (memoria.length < marcos) {
            memoria.push(pagina);
            cola_fifo.push(pagina);
            accion = `Cargar en marco ${memoria.length - 1}`;
        } else {
            const pagina_reemplazar = cola_fifo.shift();
            const idx = memoria.indexOf(pagina_reemplazar);
            memoria[idx] = pagina;
            cola_fifo.push(pagina);
            accion = `Reemplazar ${pagina_reemplazar} por ${pagina} en marco ${idx}`;
        }
    }
    // Guardar snapshot del estado de los marcos
    let snapshot = memoria.slice();
    while (snapshot.length < marcos) snapshot.push("-");
    estados_memoria.push(snapshot);
    // Mostrar estado actual (dinámico según número de marcos)
    let linea = `   ${(i+1).toString().padStart(2)}    |     ${pagina}      `;
    for (let val of snapshot) linea += `|    ${val}    `;
    linea += `|  ${fallos[i]}   | ${accion}`;
    console.log(linea);
}

// Construcción del arreglo de resultados
let resultado = [];
for (let i = 0; i < referencias.length; i++) {
    let fila = [`Instante ${i+1}`, referencias[i]];
    fila = fila.concat(estados_memoria[i]);
    fila.push(fallos[i]);
    resultado.push(fila);
}

console.log("\n=== ARREGLO DE RESULTADOS ===");
let cabecera = ["Instante", "Referencia"];
for (let i = 0; i < marcos; i++) cabecera.push(`Marco ${i}`);
cabecera.push("Fallo");
console.log(JSON.stringify(cabecera));
for (let fila of resultado) {
    console.log(JSON.stringify(fila));
}

// Mostrar resumen de resultados
const num_fallos = fallos.filter(f => f === "F").length;
const tasa_fallos = num_fallos / referencias.length;
const rendimiento = 100 * (referencias.length - num_fallos) / referencias.length;

console.log("\n=== RESULTADOS FINALES ===");
console.log(`Número de Referencias: ${referencias.length}`);
console.log(`Número de Fallos: ${num_fallos}`);
console.log(`Tasa de Fallos: ${tasa_fallos.toFixed(4)}`);
console.log(`Rendimiento (%): ${rendimiento.toFixed(2)}%`);

// Verificación de integridad
console.log("\n=== VERIFICACIÓN DE INTEGRIDAD ===");
console.log(`Estado final de memoria: ${JSON.stringify(memoria)}`);
console.log(`Estado final de cola FIFO: ${JSON.stringify(cola_fifo)}`);

// Análisis adicional
console.log("\n=== ANÁLISIS DETALLADO ===");
console.log("Secuencia de fallos por página:");
for (let i = 0; i < referencias.length; i++) {
    let estado = fallos[i] === "F" ? "FALLO" : "HIT";
    console.log(`  ${referencias[i]}: ${estado}`);
}

// ===============================
// GENERACIÓN DEL ARCHIVO CSV (como string)
// ===============================

function generarCSV() {
    let csv = '';
    // Cabeceras de estadísticas
    csv += ['Numero de fallos', num_fallos, ...Array(referencias.length-1).fill('')].join(',') + '\n';
    csv += ['Rendimiento', `${rendimiento.toFixed(0)}%`, ...Array(referencias.length-1).fill('')].join(',') + '\n';
    // Fila de referencias
    csv += ['Referencias', ...referencias].join(',') + '\n';
    // Estados de marcos
    for (let i = 0; i < marcos; i++) {
        let fila = [`Marco${i+1}`];
        for (let j = 0; j < referencias.length; j++) {
            fila.push(estados_memoria[j][i]);
        }
        csv += fila.join(',') + '\n';
    }
    // Fallos
    csv += ['FALLO', ...fallos].join(',') + '\n';
    // Reset (no aplica, pero se mantiene formato)
    csv += ['RESET', ...Array(referencias.length).fill('X')].join(',') + '\n';
    return csv;
}

const csvString = generarCSV();
console.log("\n=== CONTENIDO DEL CSV ===");
console.log(csvString);

File: fifo/fifo.py
# Simulación del algoritmo de reemplazo de páginas FIFO - Versión Corregida

# Secuencia de referencias de páginas
referencias = ["X", "Y", "Z", "W", "X", "Y", "L", "X", "Y", "Z", "W", "L"]
marcos = 4  # Número de marcos disponibles

# Inicialización
memoria = []  # Lista de páginas en memoria
cola_fifo = []  # Cola FIFO para el orden de llegada
fallos = []  # Lista de "SI" o "NO" indicando si hubo fallo de página
estados_memoria = []  # Lista para guardar el estado de los marcos en cada instante

print("=== SIMULACIÓN PASO A PASO ===")
print("Instante | Referencia | Marco 0 | Marco 1 | Marco 2 | Fallo | Acción")
print("-" * 70)

# Simulación de FIFO
for i, pagina in enumerate(referencias):
    accion = ""
    
    if pagina in memoria:
        # Página ya está en memoria - no hay fallo
        fallos.append("NO")
        accion = "Hit - página ya en memoria"
    else:
        # Página no está en memoria - hay fallo
        fallos.append("SI")
        
        if len(memoria) < marcos:
            # Hay espacio disponible
            memoria.append(pagina)
            cola_fifo.append(pagina)
            accion = f"Cargar en marco {len(memoria)-1}"
        else:
            # No hay espacio - reemplazo FIFO
            pagina_reemplazar = cola_fifo.pop(0)  # La más antigua
            
            idx = memoria.index(pagina_reemplazar)
            memoria[idx] = pagina
            cola_fifo.append(pagina)
            accion = f"Reemplazar {pagina_reemplazar} por {pagina} en marco {idx}"

    # Guardar snapshot del estado de los marcos
    snapshot = memoria.copy()
    while len(snapshot) < marcos:
        snapshot.append("-")
    estados_memoria.append(snapshot)
    
    # Mostrar estado actual
    print(f"   {i+1:2d}    |     {pagina}      |    {snapshot[0]}    |    {snapshot[1]}    |    {snapshot[2]}    |  {fallos[i]}   | {accion}")

# Construcción del arreglo de resultados
resultado = []
for i in range(len(referencias)):
    fila = [
        f"Instante {i+1}",
        referencias[i],
        estados_memoria[i][0],
        estados_memoria[i][1],
        estados_memoria[i][2],
        fallos[i]
    ]
    resultado.append(fila)

print("\n=== ARREGLO DE RESULTADOS ===")
print("['Instante', 'Referencia', 'Marco 0', 'Marco 1', 'Marco 2', 'Fallo']")
for fila in resultado:
    print(f"{fila},")

# Mostrar resumen de resultados
num_fallos = fallos.count("SI")
tasa_fallos = num_fallos / len(referencias)
rendimiento = 100 * (len(referencias) - num_fallos) / len(referencias)

print("\n=== RESULTADOS FINALES ===")
print(f"Número de Referencias: {len(referencias)}")
print(f"Número de Fallos: {num_fallos}")
print(f"Tasa de Fallos: {tasa_fallos:.4f}")
print(f"Rendimiento (%): {rendimiento:.2f}%")

# Verificación de integridad
print("\n=== VERIFICACIÓN DE INTEGRIDAD ===")
print(f"Estado final de memoria: {memoria}")
print(f"Estado final de cola FIFO: {cola_fifo}")
# Esta comparación puede ser falsa incluso si el algoritmo es correcto
# Mejor mostrar ambas listas para inspección manual

# Análisis adicional
print("\n=== ANÁLISIS DETALLADO ===")
print("Secuencia de fallos por página:")
for i, ref in enumerate(referencias):
    estado = "FALLO" if fallos[i] == "SI" else "HIT"
    print(f"  {ref}: {estado}")